api/service/explorer/storage.go

```diff
@@ -69,78 +103,239 @@ func (storage *Storage) Init(ip, port string) {
+		// flush when reached threshold
+		storage.lock.RLock()
+		numDirty := len(storage.dirtyBNs)
+		storage.lock.RUnlock()
+		if numDirty >= flushNumberThreshold || time.Since(lastFlush) >= flushTimeThreshold {
+			lastFlush = time.Now()
+			if err := storage.flushLocked(); err != nil {
+				utils.Logger().Error().Err(err).Msg("[explorer] failed to flush")
+			}
+		}
+		// prioritize processing new blocks and close C
+		select {
+		case b := <-storage.newBlockC:
+			if err := storage.handleBlock(b); err != nil {
+				utils.Logger().Error().Err(err).Msg("[explorer] handle new block result failed")
+			}
+			continue
+		case <-storage.closeC:
+			return
+		default:
+		}
+		// handles catchupBlock
+		select {
+		case b := <-storage.newBlockC:
+			if err := storage.handleBlock(b); err != nil {
+				utils.Logger().Error().Err(err).Msg("[explorer] handle new block result failed")
+			}
+		case b := <-storage.catchupBlockC:
+			if err := storage.handleBlock(b); err != nil {
+				utils.Logger().Error().Err(err).Msg("[explorer] handle catch up block failed")
+			}
+		case <-storage.closeC:
+			return
+		}
 	}
-	storage.addrIndexCache, _ = lru.New(addrIndexCacheSize)
 }
 
 // GetDB returns the LDBDatabase of the storage.
 func (storage *Storage) GetDB() *leveldb.DB {
 	return storage.db
 }
 
-// Dump extracts information from block and index them into lvdb for explorer.
-func (storage *Storage) Dump(block *types.Block, height uint64) {
-	// Skip dump for redundant blocks with lower block number than the checkpoint block number
-	blockCheckpoint := GetCheckpointKey(block.Header().Number())
-	if _, err := storage.GetDB().Get([]byte(blockCheckpoint), nil); err == nil {
+// DumpNewBlock extracts information from block and index them into leveldb for explorer.
+func (storage *Storage) DumpNewBlock(block *types.Block) {
+	if block == nil {
 		return
 	}
+	storage.lock.RLock()
+	if storage.isBlockComputed(block.NumberU64()) {
+		storage.lock.RUnlock()
+		return
+	}
+	storage.lock.RUnlock()
 
-	storage.lock.Lock()
-	defer storage.lock.Unlock()
+	storage.newBlockC <- block // blocking if channel is full
+}
+
+func (storage *Storage) DumpCatchupBlock(block *types.Block) {
 	if block == nil {
 		return
 	}
+	storage.lock.RLock()
+	if storage.isBlockComputed(block.NumberU64()) {
+		storage.lock.RUnlock()
+		return
+	}
+	storage.lock.RUnlock()
 
-	acntsTxns, acntsStakingTxns := computeAccountsTransactionsMapForBlock(block)
+	storage.catchupBlockC <- block
+}
 
-	for address, txRecords := range acntsTxns {
-		storage.UpdateTxAddressStorage(address, txRecords, false /* isStaking */)
+func (storage *Storage) isBlockComputed(bn uint64) bool {
+	blockCheckpoint := GetCheckpointKey(new(big.Int).SetUint64(bn))
+	if _, err := storage.GetDB().Get(blockCheckpoint, nil); err == nil {
+		return true
 	}
-	for address, txRecords := range acntsStakingTxns {
-		storage.UpdateTxAddressStorage(address, txRecords, true /* isStaking */)
+	if _, isDirty := storage.dirtyBNs[bn]; isDirty {
```

Comment:
shall we check on memory first to avoid potential db read?

Reply:
OK. Reasonable.